File: spec/test_tree_invalid_positions.c
#include "testparser.h"
#include <assert.h>
#include <string.h>
#include "testutils.h"

int main()
{
    char const * input = "\na\n  bb ccc";
    p_context_t * context;
    context = p_context_new((uint8_t const *)input, strlen(input));
    assert(p_parse(context) == P_SUCCESS);
    Start start = p_result(context);
    T t1 = p_Start_pT1(start);
    Token k1 = p_T_pToken(t1);
    A a1 = p_T_pA(t1);

    assert_eq(2, p_node_position(k1).row);
    assert_eq(1, p_node_position(k1).col);
    assert_eq(2, p_node_end_position(k1).row);
    assert_eq(1, p_node_end_position(k1).col);
    assert(p_position_valid(p_node_position(a1)));
    assert_eq(3, p_node_position(a1).row);
    assert_eq(3, p_node_position(a1).col);
    assert_eq(3, p_node_end_position(a1).row);
    assert_eq(8, p_node_end_position(a1).col);
    assert_eq(2, p_node_position(t1).row);
    assert_eq(1, p_node_position(t1).col);
    assert_eq(3, p_node_end_position(t1).row);
    assert_eq(8, p_node_end_position(t1).col);

    assert_eq(2, p_node_position(start).row);
    assert_eq(1, p_node_position(start).col);
    assert_eq(3, p_node_end_position(start).row);
    assert_eq(8, p_node_end_position(start).col);

    p_context_delete(context);

    input = "a\nbb";
    context = p_context_new((uint8_t const *)input, strlen(input));
    assert(p_parse(context) == P_SUCCESS);
    start = p_result(context);
    t1 = p_Start_pT1(start);
    k1 = p_T_pToken(t1);
    a1 = p_T_pA(t1);

    assert_eq(1, p_node_position(k1).row);
    assert_eq(1, p_node_position(k1).col);
    assert_eq(1, p_node_end_position(k1).row);
    assert_eq(1, p_node_end_position(k1).col);
    assert(p_position_valid(p_node_position(a1)));
    assert_eq(2, p_node_position(a1).row);
    assert_eq(1, p_node_position(a1).col);
    assert_eq(2, p_node_end_position(a1).row);
    assert_eq(2, p_node_end_position(a1).col);
    assert_eq(1, p_node_position(t1).row);
    assert_eq(1, p_node_position(t1).col);
    assert_eq(2, p_node_end_position(t1).row);
    assert_eq(2, p_node_end_position(t1).col);

    assert_eq(1, p_node_position(start).row);
    assert_eq(1, p_node_position(start).col);
    assert_eq(2, p_node_end_position(start).row);
    assert_eq(2, p_node_end_position(start).col);

    p_context_delete(context);

    input = "a\nc\nc";
    context = p_context_new((uint8_t const *)input, strlen(input));
    assert(p_parse(context) == P_SUCCESS);
    start = p_result(context);
    t1 = p_Start_pT1(start);
    k1 = p_T_pToken(t1);
    a1 = p_T_pA(t1);

    assert_eq(1, p_node_position(k1).row);
    assert_eq(1, p_node_position(k1).col);
    assert_eq(1, p_node_end_position(k1).row);
    assert_eq(1, p_node_end_position(k1).col);
    assert(p_position_valid(p_node_position(a1)));
    assert_eq(2, p_node_position(a1).row);
    assert_eq(1, p_node_position(a1).col);
    assert_eq(3, p_node_end_position(a1).row);
    assert_eq(1, p_node_end_position(a1).col);
    assert_eq(1, p_node_position(t1).row);
    assert_eq(1, p_node_position(t1).col);
    assert_eq(3, p_node_end_position(t1).row);
    assert_eq(1, p_node_end_position(t1).col);

    assert_eq(1, p_node_position(start).row);
    assert_eq(1, p_node_position(start).col);
    assert_eq(3, p_node_end_position(start).row);
    assert_eq(1, p_node_end_position(start).col);

    p_context_delete(context);

    input = "a";
    context = p_context_new((uint8_t const *)input, strlen(input));
    assert(p_parse(context) == P_SUCCESS);
    start = p_result(context);
    t1 = p_Start_pT1(start);
    k1 = p_T_pToken(t1);
    a1 = p_T_pA(t1);

    assert_eq(1, p_node_position(k1).row);
    assert_eq(1, p_node_position(k1).col);
    assert_eq(1, p_node_end_position(k1).row);
    assert_eq(1, p_node_end_position(k1).col);
    assert(!p_position_valid(p_node_position(a1)));
    assert_eq(1, p_node_position(t1).row);
    assert_eq(1, p_node_position(t1).col);
    assert_eq(1, p_node_end_position(t1).row);
    assert_eq(1, p_node_end_position(t1).col);

    assert_eq(1, p_node_position(start).row);
    assert_eq(1, p_node_position(start).col);
    assert_eq(1, p_node_end_position(start).row);
    assert_eq(1, p_node_end_position(start).col);

    p_context_delete(context);

    return 0;
}

File: spec/test_tree_invalid_positions.rs
use testparser::*;

fn main() {
    let mut context = p_context_new(b"\na\n  bb ccc");
    assert_eq!(P_SUCCESS, p_parse(&mut context));
    {
        let start = p_result(&context);
        let t1 = start.pT1();
        let k1 = t1.pToken();
        let a1 = t1.pA();
        assert_eq!(2, k1.position().row);
        assert_eq!(1, k1.position().col);
        assert_eq!(2, k1.end_position().row);
        assert_eq!(1, k1.end_position().col);
        assert!(a1.position().valid());
        assert_eq!(3, a1.position().row);
        assert_eq!(3, a1.position().col);
        assert_eq!(3, a1.end_position().row);
        assert_eq!(8, a1.end_position().col);
        assert_eq!(2, t1.position().row);
        assert_eq!(1, t1.position().col);
        assert_eq!(3, t1.end_position().row);
        assert_eq!(8, t1.end_position().col);
        assert_eq!(2, start.position().row);
        assert_eq!(1, start.position().col);
        assert_eq!(3, start.end_position().row);
        assert_eq!(8, start.end_position().col);
    }
    p_context_delete(context);

    let mut context = p_context_new(b"a\nbb");
    assert_eq!(P_SUCCESS, p_parse(&mut context));
    {
        let start = p_result(&context);
        let t1 = start.pT1();
        let k1 = t1.pToken();
        let a1 = t1.pA();
        assert_eq!(1, k1.position().row);
        assert_eq!(1, k1.position().col);
        assert!(a1.position().valid());
        assert_eq!(2, a1.position().row);
        assert_eq!(1, a1.position().col);
        assert_eq!(2, a1.end_position().row);
        assert_eq!(2, a1.end_position().col);
        assert_eq!(1, t1.position().row);
        assert_eq!(2, t1.end_position().row);
        assert_eq!(2, t1.end_position().col);
        assert_eq!(1, start.position().row);
        assert_eq!(2, start.end_position().row);
        assert_eq!(2, start.end_position().col);
    }
    p_context_delete(context);

    let mut context = p_context_new(b"a\nc\nc");
    assert_eq!(P_SUCCESS, p_parse(&mut context));
    {
        let start = p_result(&context);
        let t1 = start.pT1();
        let a1 = t1.pA();
        assert!(a1.position().valid());
        assert_eq!(2, a1.position().row);
        assert_eq!(1, a1.position().col);
        assert_eq!(3, a1.end_position().row);
        assert_eq!(1, a1.end_position().col);
        assert_eq!(1, t1.position().row);
        assert_eq!(3, t1.end_position().row);
        assert_eq!(1, t1.end_position().col);
        assert_eq!(3, start.end_position().row);
        assert_eq!(1, start.end_position().col);
    }
    p_context_delete(context);

    let mut context = p_context_new(b"a");
    assert_eq!(P_SUCCESS, p_parse(&mut context));
    {
        let start = p_result(&context);
        let t1 = start.pT1();
        let a1 = t1.pA();
        assert!(!a1.position().valid());
        assert_eq!(1, t1.position().row);
        assert_eq!(1, t1.position().col);
        assert_eq!(1, t1.end_position().row);
        assert_eq!(1, t1.end_position().col);
        assert_eq!(1, start.position().row);
        assert_eq!(1, start.end_position().row);
        assert_eq!(1, start.end_position().col);
    }
    p_context_delete(context);
}
